### PR TITLE
add support for sand ortholinear keyboard

### PR DESCRIPTION
**Pull request by Sandpiper131**
Add support for a 47 key ortholinear keyboard I designed

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/sandortho/sandortho.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    return process_record_user(keycode, record);
+}
+
+bool led_update_kb(led_t led_state) {
+    // put your keyboard LED indicator (ex: Caps Lock LED) toggling code here
+
+    return led_update_user(led_state);
+}
+*/
```
Comment (drashna) [inline keyboards/sandortho/sandortho.c:55]: These aren't used, and can be removed.
```suggestion
```

File: keyboards/sandortho/keymaps/default/keymap.c [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    KC_NO,     KC_LCTRL, KC_LALT, KC_LGUI, MO(_LOWER), KC_NO,     KC_SPACE,     MO(_RAISE), KC_LEFT,  KC_DOWN,  KC_UP,       KC_RIGHT  //HC_NO before space is not a key in hardware
+    ),
+    [_RAISE] = LAYOUT(
+   	KC_NO,     KC_1,     KC_2,    KC_3,    KC_4,       KC_5,      KC_6,         KC_7,       KC_8,     KC_9,      KC_0,       KC_DELETE,               
+   	KC_NO,     KC_4,     KC_5,    KC_6,    KC_0,       KC_NO,     KC_NO,        KC_NO,      KC_NO,    KC_MINUS,  KC_EQUAL,   KC_NO,  
+   	KC_LSHIFT, KC_7,     KC_8,    KC_9,    KC_NO,      KC_NO,     KC_NO,        KC_NO,      KC_NO,    KC_NO,     KC_NO,      KC_NO,  
+   	RESET,     KC_LCTRL, KC_LALT, KC_LGUI, KC_NO,      KC_NO,     KC_SPACE,     KC_NO,      KC_NO,    KC_NO,     KC_NO,      KC_NO
+    ),
+    [_LOWER] = LAYOUT(
+   	KC_NO,     EXCL,     AT,      POUND,   DOLLAR,     PERCENT,   CARAT,        AND,        STAR,     LPAR,      RPAR,        KC_DELETE,
```
Comment (drashna) [inline keyboards/sandortho/keymaps/default/keymap.c:57]:
```suggestion
   	KC_NO,     KC_EXCL,     KC_AT,      KC_HASH,   KC_DLR,     KC_PERCENT,   KC_CIRC,        KC_AMP,        KC_ASTR,     KC_LPRN,      KC_RPRN,        KC_DELETE,               
```

File: keyboards/sandortho/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	
+	    }
+	    break;
+	case RPAR:
+	    if (record->event.pressed) {
+	    	SEND_STRING(")");
+	    }else{
+	
+	    }
+	    break;
```
Comment (drashna) [inline keyboards/sandortho/keymaps/default/keymap.c:157]: Just a heads up, most of these actually have dedicated keycodes. 

https://docs.qmk.fm/#/keycodes?id=us-ansi-shifted-symbols

And for the most part, they're just `LSFT(kc)`, actually.

```suggestion
```

File: keyboards/sandortho/keymaps/default/config.h [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/* Copyright 2020 Luke LaCasse
```
Comment (zvecr) [inline keyboards/sandortho/keymaps/default/config.h:1]: This file should be deleted

File: keyboards/sandortho/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+void matrix_scan_user(void) {
+
+}
+
+bool led_update_user(led_t led_state) {
+    return true;
+}
+*/
```
Comment (zvecr) [inline keyboards/sandortho/keymaps/default/keymap.c:174]:
```suggestion
```

File: keyboards/sandortho/readme.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# sandortho
+
+![sandortho](imgur.com image replace me!)
```
Comment (zvecr) [inline keyboards/sandortho/readme.md:3]: This should be filled in, or removed

File: keyboards/sandortho/sandortho.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k110, k111, \
+        k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k210, k211, \
+	  k30, k31, k32, k33, k34, k35, k36, k37, k38, k39, k310, k311 \
+) \
+{ \
+    { k00, k01, k02, k03, k04, k05, k06, k07, k08, k09, k010, k011}, \
+    { k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k110, k111},  \
+    { k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k210, k211}, \
+    { k30, k31, k32, k33, k34, k35, k36, k37, k38, k39, k310, k311} \
+}
```
Comment (zvecr) [inline keyboards/sandortho/sandortho.h:40]: This abstract physical layout, and not require the end user to add a bunch of KC_NO in dead spots
```suggestion
#define LAYOUT( \
    k00, k01, k02, k03, k04, k05, k06, k07, k08, k09, k010, k011, \
      k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k110, k111, \
        k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k210, k211, \
	  k30, k31, k32, k33, k34,   k36,   k37, k38, k39, k310, k311 \
) \
{ \
    { k00, k01, k02, k03, k04, k05, k06, k07, k08, k09, k010, k011}, \
    { k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k110, k111},  \
    { k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k210, k211}, \
    { k30, k31, k32, k33, k34, KC_NO, k36, k37, k38, k39, k310, k311} \
}
```

File: keyboards/sandortho/sandortho.c (hunk 1 above)
Comment (zvecr) [inline keyboards/sandortho/sandortho.c:55]:
```suggestion
```

File: keyboards/sandortho/rules.mk [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+COMMAND_ENABLE = no        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output
+
```
Comment (zvecr) [inline keyboards/sandortho/rules.mk:30]:
```suggestion
# MCU name
MCU = atmega32u4

# Bootloader selection
BOOTLOADER = atmel-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/sandortho/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    RPAR
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    /* Base */
+    [_BASE] = LAYOUT(
+	KC_TAB,    KC_Q,     KC_W,    KC_E,    KC_R,       KC_T,      KC_Y,         KC_U,       KC_I,     KC_O,     KC_P,        KC_BSPACE,
+	KC_ESCAPE, KC_A,     KC_S,    KC_D,    KC_F,       KC_G,      KC_H,         KC_J,       KC_K,     KC_L,     KC_SCOLON,   KC_QUOTE,
+	KC_LSHIFT, KC_Z,     KC_X,    KC_C,    KC_V,       KC_B,      KC_N,         KC_M,       KC_COMMA, KC_DOT,   KC_SLASH,    KC_ENTER,
+    KC_NO,     KC_LCTRL, KC_LALT, KC_LGUI, MO(_LOWER), KC_NO,     KC_SPACE,     MO(_RAISE), KC_LEFT,  KC_DOWN,  KC_UP,       KC_RIGHT  //HC_NO before space is not a key in hardware
```
Comment (zvecr) [inline keyboards/sandortho/keymaps/default/keymap.c:48]: Same for the rest of the layers, these should be updated to the layout macro
```suggestion
    KC_NO,     KC_LCTRL, KC_LALT, KC_LGUI, MO(_LOWER), KC_SPACE,     MO(_RAISE), KC_LEFT,  KC_DOWN,  KC_UP,       KC_RIGHT
```

File: keyboards/sandortho/config.h [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+#define LCD_RW_PORT      LCD_PORT     //< port for RW line
+#define LCD_RW_PIN       2            //< pin  for RW line
+#define LCD_E_PORT       LCD_PORT     //< port for Enable line
+#define LCD_E_PIN        1            //< pin  for Enable line
+#endif
+*/
+
+/* Bootmagic Lite key configuration */
+// #define BOOTMAGIC_LITE_ROW 0
+// #define BOOTMAGIC_LITE_COLUMN 0
```
Comment (zvecr) [inline keyboards/sandortho/config.h:250]:
```suggestion
/*
 * Feature disable options
 *  These options are also useful to firmware size reduction.
 */

/* disable debug print */
//#define NO_DEBUG

/* disable print */
//#define NO_PRINT

/* disable action features */
//#define NO_ACTION_LAYER
//#define NO_ACTION_TAPPING
//#define NO_ACTION_ONESHOT

/* disable these deprecated features by default */
#define NO_ACTION_MACRO
#define NO_ACTION_FUNCTION

/* Bootmagic Lite key configuration */
//#define BOOTMAGIC_LITE_ROW 0
//#define BOOTMAGIC_LITE_COLUMN 0
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/sandortho/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            if (record->event.pressed) {
+                // when keycode QMKURL is pressed
+                SEND_STRING("https://qmk.fm/\n");
+            } else {
+                // when keycode QMKURL is released
+            }
+            break;
+    }
+    return true;
+}
```
Comment (noroadsleft) [inline keyboards/sandortho/keymaps/default/keymap.c:90]: Unused.

```suggestion
```

File: keyboards/sandortho/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+   	KC_NO,     KC_4,     KC_5,    KC_6,    KC_0,       KC_NO,     KC_NO,        KC_NO,      KC_NO,    KC_MINUS,  KC_EQUAL,   KC_NO,  
+   	KC_LSHIFT, KC_7,     KC_8,    KC_9,    KC_NO,      KC_NO,     KC_NO,        KC_NO,      KC_NO,    KC_NO,     KC_NO,      KC_NO,  
+   	RESET,     KC_LCTRL, KC_LALT, KC_LGUI, KC_NO,      KC_NO,     KC_SPACE,     KC_NO,      KC_NO,    KC_NO,     KC_NO,      KC_NO
+    ),
+    [_LOWER] = LAYOUT(
+   	KC_NO,     KC_EXCL,     KC_AT,      KC_HASH,   KC_DLR,     KC_PERCENT,   KC_CIRC,        KC_AMP,        KC_ASTR,     KC_LPRN,      KC_RPRN,        KC_DELETE,               
+   	KC_NO,     KC_F1,    KC_F2,   KC_F3,   KC_F4,      KC_F5,     KC_F6,        KC_F7,      KC_F8,    KC_BSLASH, KC_LBRACKET, KC_RBRACKET,  
+   	KC_LSHIFT, KC_F9,    KC_F10,  KC_F11,  KC_F12,     KC_NO,     KC_NO,        KC_NO,      KC_NO,    KC_NO,     KC_NO,       KC_NO,  
+   	KC_NO,     KC_LCTRL, KC_LALT, KC_LGUI, KC_NO,      KC_NO,     KC_SPACE,     KC_NO,      KC_HOME,  KC_PGDOWN, KC_PGUP,     KC_END
+    )/*,
```
Comment (noroadsleft) [inline keyboards/sandortho/keymaps/default/keymap.c:61]: - use short-form keycodes to make this easier to read
- remove an extra keycode from the `_RAISE` and `_LOWER` layers that prevented the keymap from being compiled

```suggestion
    [_BASE] = LAYOUT(
        KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,       KC_T,    KC_Y,    KC_U,       KC_I,    KC_O,    KC_P,    KC_BSPC,
        KC_ESC,  KC_A,    KC_S,    KC_D,    KC_F,       KC_G,    KC_H,    KC_J,       KC_K,    KC_L,    KC_SCLN, KC_QUOT,
        KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,       KC_B,    KC_N,    KC_M,       KC_COMM, KC_DOT,  KC_SLSH, KC_ENT,
        KC_NO,   KC_LCTL, KC_LALT, KC_LGUI, MO(_LOWER),     KC_SPC,       MO(_RAISE), KC_LEFT, KC_DOWN, KC_UP,   KC_RGHT
    ),
    [_RAISE] = LAYOUT(
        KC_NO,   KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_DEL,
        KC_NO,   KC_4,    KC_5,    KC_6,    KC_0,    KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_MINS, KC_EQL,  KC_NO,
        KC_LSFT, KC_7,    KC_8,    KC_9,    KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,
        RESET,   KC_LCTL, KC_LALT, KC_LGUI, KC_NO,       KC_SPC,       KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO
    ),
    [_LOWER] = LAYOUT(
        KC_NO,   KC_EXCL, KC_AT,   KC_HASH, KC_DLR,  KC_PERC, KC_CIRC, KC_AMP,  KC_ASTR, KC_LPRN, KC_RPRN, KC_DEL,
        KC_NO,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_BSLS, KC_LBRC, KC_RBRC,
        KC_LSFT, KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,   KC_NO,
        KC_NO,   KC_LCTL, KC_LALT, KC_LGUI, KC_NO,       KC_SPC,       KC_NO,   KC_HOME, KC_PGDN, KC_PGUP, KC_END
    )/*,
```

File: keyboards/sandortho/readme.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+The sand ortholinear keyboard was inpired by silicon, the element used to make semiconductor materials. The keyboard has 47 keys laid out in an ortholinear grid with a 2U spacebar. The board can be assembled with standoffs in between the backplate, the keyboard can be as short as 7.2mm tall (excluding switches and keycpas) or as tall as ~9.5mm with a 1/4" piece of acrylic sandwiched in between. I worked on this project from the end of July 2020 to September 2020.
```
Comment (noroadsleft) [inline keyboards/sandortho/readme.md:5]: Typo fixes.

```suggestion
The sand ortholinear keyboard was inspired by silicon, the element used to make semiconductor materials. The keyboard has 47 keys laid out in an ortholinear grid with a 2U spacebar. The board can be assembled with standoffs in between the backplate, the keyboard can be as short as 7.2mm tall (excluding switches and keycaps) or as tall as ~9.5mm with a 1/4" piece of acrylic sandwiched in between. I worked on this project from the end of July 2020 to September 2020.
```

File: keyboards/sandortho/readme.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+* Keyboard Maintainer: [Luke LaCasse](https://github.com/Sandpiper131)
+* Hardware Supported: ATMEGA32-U4
```
Comment (noroadsleft) [inline keyboards/sandortho/readme.md:8]:
```suggestion
* Hardware Supported: ATmega32U4
```

File: keyboards/sandortho/sandortho.h [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+#define LAYOUT( \
+    k00, k01, k02, k03, k04, k05, k06, k07, k08, k09, k010, k011, \
+      k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k110, k111, \
+        k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k210, k211, \
+	  k30, k31, k32, k33, k34,   k36,   k37, k38, k39, k310, k311 \
```
Comment (noroadsleft) [inline keyboards/sandortho/sandortho.h:33]: Cleans up the spacing/indentation.

```suggestion
    k10, k11, k12, k13, k14, k15, k16, k17, k18, k19, k110, k111, \
    k20, k21, k22, k23, k24, k25, k26, k27, k28, k29, k210, k211, \
    k30, k31, k32, k33, k34,   k36,    k37, k38, k39, k310, k311 \
```

File: keyboards/sandortho/info.json [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+                {"label":"k10", "x":0, "y":1},{"label":"k11", "x":1, "y":1},{"label":"k12", "x":2, "y":1},{"label":"k13", "x":3, "y":1},{"label":"k14", "x":4, "y":1},{"label":"k15", "x":5, "y":1},
+                {"label":"k16", "x":6, "y":1},{"label":"k17", "x":7, "y":1},{"label":"k18", "x":8, "y":1},{"label":"k19", "x":9, "y":1},{"label":"k110", "x":10, "y":1},{"label":"k111", "x":11, "y":1},
+                {"label":"k20", "x":0, "y":2},{"label":"k21", "x":1, "y":2},{"label":"k22", "x":2, "y":2},{"label":"k23", "x":3, "y":2},{"label":"k24", "x":4, "y":2},{"label":"k25", "x":5, "y":2},
+                {"label":"k26", "x":6, "y":2},{"label":"k27", "x":7, "y":2},{"label":"k28", "x":8, "y":2},{"label":"k29", "x":9, "y":2},{"label":"k210", "x":10, "y":2},{"label":"k211", "x":11, "y":2},
+                {"label":"k30", "x":0, "y":3},{"label":"k31", "x":1, "y":3},{"label":"k32", "x":2, "y":3},{"label":"k33", "x":3, "y":3},{"label":"k34", "x":4, "y":3},{"label":"k35", "x":5, "y":3},
+                {"label":"k36", "x":6, "y":3},{"label":"k37", "x":7, "y":3},{"label":"k38", "x":8, "y":3},{"label":"k39", "x":9, "y":3},{"label":"k310", "x":10, "y":3},{"label":"k311", "x":11, "y":3}
```
Comment (noroadsleft) [inline keyboards/sandortho/info.json:17]: Bugfix for an extra object.

```suggestion
                {"label":"k30", "x":0, "y":3},{"label":"k31", "x":1, "y":3},{"label":"k32", "x":2, "y":3},{"label":"k33", "x":3, "y":3},{"label":"k34", "x":4, "y":3},
                {"label":"k36", "x":5, "y":3, "w":2},{"label":"k37", "x":7, "y":3},{"label":"k38", "x":8, "y":3},{"label":"k39", "x":9, "y":3},{"label":"k310", "x":10, "y":3},{"label":"k311", "x":11, "y":3}
```

File: keyboards/sandortho/config.h [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+//   #define RGBLIGHT_EFFECT_RGB_TEST
+//   #define RGBLIGHT_EFFECT_ALTERNATING
+// /*== customize breathing effect ==*/
+//   /*==== (DEFAULT) use fixed table instead of exp() and sin() ====*/
+//   #define RGBLIGHT_BREATHE_TABLE_SIZE 256      // 256(default) or 128 or 64
+//   /*==== use exp() and sin() ====*/
+//   #define RGBLIGHT_EFFECT_BREATHE_CENTER 1.85  // 1 to 2.7
+//   #define RGBLIGHT_EFFECT_BREATHE_MAX    255   // 0 to 255
+// #endif
+
```
Comment (noroadsleft) [inline keyboards/sandortho/config.h:86]: None of this is needed.

```suggestion
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]
GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/10543/files).

**Pull request closed without merging**